### ISSUE TITLE
Retry Stellar transactions indefinitely

### ISSUE THREAD

**Issue by ebma** [pendulum-chain/spacewalk#315]
We recently changed the logic in the `execute_open_requests()` function to retry up to some specific number of times (ie `3`). Now, while this improves our resilience to some network delays, we should actually make the vault client retry the submission indefinitely because it might happen that an attacker manages to spam/delay the vault's account for a longer timeframe, thus making the limited number of retries problematic.
I would suggest that we extend the [task](https://github.com/pendulum-chain/spacewalk/blob/main/clients/wallet/src/task.rs) implementation to also be able to retry indefinitely, (as long as recoverable errors are encountered).

**Comment by ebma** [comment]
But an estimate of 21 would mean that this task is so huge that it should be split up into multiple tasks 😅 . I'll invite for planning poker

**Comment by b-yap** [comment]
okay, so maybe it's an 8. I'll update my estimate.

Currently, I'm listing down what "errors" are deemed acceptable and what are "not". And how do we communicate it back to the user. Especially when resubmitting stellar transactions that were in the cache, how do we let the user know that a few of them failed? 

For recoverable errors, let's say (stellar transactions) tx1, tx2, ... until tx10: if at tx6 it fails repeatedly (do you think it can happen, @ebma , @TorstenStueber ?) that means tx7 and beyond will have to wait indefinitely too.

**Comment by b-yap** [comment]
```rust
pub enum Error {
	#[error("Invalid secret key")]
	InvalidSecretKey,
	#[error("Error fetching horizon data: {0}")]
	HttpFetchingError(#[from] FetchError),
	#[error("Could not build transaction: {0}")]
	BuildTransactionError(String),
	#[error("Transaction submission failed. Title: {title}, Status: {status}, Reason: {reason}, Envelope XDR: {envelope_xdr:?}")]
	HorizonSubmissionError {
		title: String,
		status: u16,
		reason: String,
		envelope_xdr: Option<String>,
	},
	#[error("Could not parse string: {0}")]
	Utf8Error(#[from] std::str::Utf8Error),
	#[error("Could not decode XDR")]
	DecodeError,
	#[error("Could not sign envelope")]
	SignEnvelopeError,

	#[error(transparent)]
	CacheError(CacheError),
}
```
This is the error returned from `submit_transaction(...)` 
@ebma 

1. For `HttpFetchingError`,  reqwest's Error  has an [`is_timeout()`](https://docs.rs/reqwest/latest/reqwest/struct.Error.html#method.is_timeout) method
2. For `HorizonSubmissionError`, if the status is [504](https://developers.stellar.org/api/errors/http-status-codes/horizon-specific/timeout#:~:text=The%20timeout%20error%20returns%20a,in%20a%20timely%20manner%2C%20or)
3. `CacheError`: 
     ```rust 
     pub enum CacheErrorKind {
	  #[error("Could not extract sequence number from transaction")]
	  UnknownSequenceNumber,
  
	  #[error("Sequence number already in use.")]
	  SequenceNumberAlreadyUsed,
  
	  #[error("Failed to create directory")]
	  CreateDirectoryFailed,
  
	  #[error("Failed to create file")]
	  FileCreationFailed,
  
	  #[error("Failed to write file")]
	  WriteToFileFailed,
  
	  #[error("Failed to delete file")]
	  DeleteFileFailed,
  
	  #[error("File does not exist")]
	  FileDoesNotExist,
  
	  #[error("Failed to read file")]
	  ReadFileFailed,
  
	  #[error("File is invalid")]
	  InvalidFile,
  
	  #[error("Failed to decode file")]
	  DecodeFileFailed,
     }
     ```
     We'll probably accept just: `CreateDirectoryFailed`, `WriteToFileFailed`, `DeleteFileFailed`


I could add an `is_recoverable()` method for the error, and checks self whether or not it's a recoverable error

**Comment by ebma** [comment]
> I could add an is_recoverable() method for the error, and checks self whether or not it's a recoverable error

Sounds good.


> For HorizonSubmissionError, if the status is [504](https://developers.stellar.org/api/errors/http-status-codes/horizon-specific/timeout#:~:text=The%20timeout%20error%20returns%20a,in%20a%20timely%20manner%2C%20or). But we won't accept [the internal server error 500](https://developers.stellar.org/api/errors/http-status-codes/standard) and [tx_internal_error](https://developers.stellar.org/api/errors/result-codes/transactions), right? If the server's at fault, we have to stop entirely.

Hmm, I am not sure actually. I think this is a case where we should still retry indefinitely. We don't know when this server works correctly again, so we can just retry submission every xx seconds and once it's running again we're automatically back in business without extra logic. The (better) alternative would be to fall back to other horizon servers. Other companies like LOBSTR or stellarx also operate horizon servers AFAIK so we could also try submitting the transaction to their servers instead. How much effort would you estimate for implementing such a fallback mechanism @b-yap? We could fallback in case we encounter any 5xx error I think. But I am worried that this is nasty to implement.

**Comment by b-yap** [comment]
A fallback is a good idea. We can have it confined in `fn submit_transaction()`. 

:sweat_smile: We do have diverging opinions with "indefinitely", for server errors. If we stop the vault with a clear error, the user will understand that it only needs to rerun once the server is up and running again. Say if the server is down for 30 minutes, the user will wonder why his vault is not doing anything else in the past few minutes, except resubmitting 1 transaction. 

If it gets stuck in `send_payment_to_address`, the user won't understand why it isn't able to execute anything new. Even if we throw an error to the user like "_still executing an existing transaction, can't accept any new ones_", it'll be tiresome after a few minutes and wonder what's wrong with the previous transaction because it's taking too long. But if the user immediately _knows_ that the server is down, isn't it better that he can conclude immediately that nothing can be done at this point and just wait for it? Or maybe for him the best way is to report back to stellar and say "hey, something's wrong with your server"?

**Comment by TorstenStueber** [comment]
For 504 errors: I think we can assume that they are transient and that submitting the transaction will _eventually_ work (that means that we need to retry indefinitely).

**Comment by ebma** [comment]
> transactions can be sent out of order without problems. If the sequence number is 7, then tx7 will not be executed yet but stays in the transaction pool. Only once tx6 reaches the transaction pool, it will then be executed and after that tx7 will be executed (possibly even within the same block, depending on how the transaction selection logic works for the node that creates the block).

That would actually be great if that's how it works. So you're saying that the transaction would only get executed if the sequence number is exactly `account_sequence_number + 1` and not just larger than the current account sequence number?

**Comment by TorstenStueber** [comment]
@ebma see [here](https://github.com/stellar/stellar-core/blob/bab0caf7bd93fc8b21402d10baac10c2564851b4/src/transactions/readme.md#sequence-number):
> The transaction number has to be the next sequence number based off the one stored in the source account when the transaction is applied.

**Comment by TorstenStueber** [comment]
The only way to step over sequence numbers is to explicitly execute a [bump sequence operation](https://developers.stellar.org/docs/fundamentals-and-concepts/list-of-operations#bump-sequence).

**Comment by b-yap** [comment]
504 error is a timeout error. Yes, I'll retry them until it succeeds (or throws a different error).

>. If the sequence number is 7, then tx7 will not be executed yet but stays in the transaction pool.  

  Cool, so we actually do not need to wait for previous transactions to finish before sending another one. That is awesome.

@ebma @TorstenStueber okay, I have had my say about server errors. I need a verdict :man_judge:
Should we keep submitting indefinitely, too?

**Comment by TorstenStueber** [comment]
@b-yap unfortunately not, and I just tested it on [Stellar laboratory](https://laboratory.stellar.org/#?network=test).

I remember that it works but probably the out of order transaction need to arrive in the pool at similar times. If you first send the transaction supposed to be executed second (with the higher sequence number) and then wait, say, 10 seconds, then it will fail. But you can just resend and it and it will eventually work (when the other one has been submitted in the mean time). 

So what I mean is that unsuccessfully sending the transaction with the higher sequence number will not lead to problems for the transaction with a lower sequence number – that one will just work once submitted and after that also the transaction with the higher number will work once resubmitted. So all out-of-order problems are just transient when trying resending...

(I just experimented with this and confirmed on Stellar [Stellar laboratory](https://laboratory.stellar.org/#?network=test)

**Comment by ebma** [comment]
Okay so if I understand correctly, we can conclude that we can keep re-submitting all transactions that throw a recoverable error forever, and the order of resubmission does not matter. So in theory @b-yap you could spawn a new task for each transaction submission because every submission can run in parallel. Maybe that's easier to implement because then you can make each task sleep for some time after each retried submission etc. (You don't have to implement it that way though, of course)

**Comment by TorstenStueber** [comment]
We should of course only resubmit when the problem is that the sequence number is not "ready yet". Unfortunately, the error returned is just "bad seq no" (or similar) and it is not clear whether the sequence number is too early or too late. If the service doesn't have logical bugs, then "bad seq no" should only happen because the transaction is too early...
